lint/linter.py

```diff
@@ -438,159 +435,82 @@ def get_view_settings(self):
-                    # extract the root folder of the currently watched file
-                    filename = view.file_name() or 'FILE NOT ON DISK'
-                    if folder['path'] in filename:
-                        expressions.append({
-                            'token': '${project}',
-                            'value': folder['path']
-                        })
-
-            expressions.append({
-                'token': '${directory}',
-                'value': directory
-            })
+                return value
 
-            expressions.append({
-                'token': '${root}',
-                'value': window_vars.get('folder', None) or directory
-            })
+        window = self.view.window()
+        variables = ChainMap(
+            {}, window.extract_variables() if window else {}, os.environ)
 
-        expressions.append({
-            'token': '${home}',
-            'value': os.path.expanduser('~').rstrip(os.sep).rstrip(os.altsep).replace('\\', '/') or 'HOME NOT SET'
-        })
+        filename = self.view.file_name()
+        project_folder = self._guess_project_path(window, filename)
+        if project_folder:
+            variables['folder'] = project_folder
 
-        expressions.append({
-            'token': '${sublime}',
-            'value': sublime.packages_path()
-        })
+        if persist.debug_mode():
+            import pprint
+            self._debug_print_available_variables(pprint.pformat(dict(variables), indent=2))
+        return recursive_replace(variables, settings)
 
-        expressions.append({
-            'token': re.compile(r'\${env:(?P<variable>[^}]+)}'),
-            'value': (
-                lambda m: os.getenv(m.group('variable')) if
-                os.getenv(m.group('variable')) else
-                "%s NOT SET" % m.group('variable'))
-        })
+    @staticmethod
+    @lru_cache(maxsize=1)
+    def _debug_print_available_variables(variables):
+        persist.debug('Available variables: {}'.format(variables))
 
-        recursive_replace(expressions, settings)
+    @staticmethod
+    def _guess_project_path(window, filename):
+        if not window:
+            return
 
-    def merge_project_settings(self, view_settings, project_settings):
-        """
-        Return this linter's view settings merged with the current project settings.
+        folders = window.folders()
+        if not folders:
+            return
 
-        Subclasses may override this if they wish to do something more than
-        replace view settings with inline settings of the same name.
-        The settings object may be changed in place.
+        if not filename:
+            return folders[0]
 
-        """
-        view_settings.update(project_settings)
-        return view_settings
+        for folder in folders:
+            # Take the first one; should we take the deepest one? The shortest?
+            if folder in filename:
+                return folder
```

Comment:
This 'algorithm' can be improved. But it is not necessary for the first iteration. 

The question here is: if you have a file open, and your *window* has multiple folders open, and more than one open folder contains the open file, which folder should be regarded as the 'project' folder. The deepest, the shortest, or the first?

Reply:
E.g. your main aka 'first' folder is the root of your current webapp. The second folder is a npm module/dependency you control. You open and edit a file from your (own) dependency. Which folder is the project folder here?

Reply:
FWIW SublimeText uses the first folder for "reveal in sidebar". I think it makes sense, or is at least easy to explain, to use the first folder as primary / most important.